File: python/example/samples/custom.py
"""
File: custom.py
Project: samples
Created Date: 11/10/2021
Author: Shun Suzuki
-----
Last Modified: 21/09/2023
Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
-----
Copyright (c) 2022-2023 Shun Suzuki. All rights reserved.

"""


import numpy as np
from numpy.typing import ArrayLike

from pyautd3 import Controller, Drive, EmitIntensity, Geometry, Phase, SamplingConfiguration, Silencer
from pyautd3.gain import Gain
from pyautd3.modulation import Modulation


class Focus(Gain):
    def __init__(self: "Focus", point: ArrayLike) -> None:
        self.point = np.array(point)

    def calc(self: "Focus", geometry: Geometry) -> dict[int, np.ndarray]:
        return Gain._transform(
            geometry,
            lambda dev, tr: Drive(
                Phase.from_rad(float(np.linalg.norm(tr.position - self.point)) * tr.wavenumber(dev.sound_speed)),
                EmitIntensity.maximum(),
            ),
        )


class Burst(Modulation):
    _length: int

    def __init__(self: "Burst", length: int, config: SamplingConfiguration | None = None) -> None:
        super().__init__(config if config is not None else SamplingConfiguration.from_frequency(4e3))
        self._length = length

    def calc(self: "Burst") -> np.ndarray:
        buf = np.array([EmitIntensity.minimum()] * self._length)
        buf[0] = EmitIntensity.maximum()
        return buf


async def custom(autd: Controller) -> None:
    config = Silencer()
    await autd.send_async(config)

    f = Focus(autd.geometry.center + np.array([0.0, 0.0, 150.0]))
    m = Burst(4000)

    await autd.send_async(m, f)

File: cs/example/Samples/Custom.cs
/*
 * File: Custom.cs
 * Project: Test
 * Created Date: 14/10/2021
 * Author: Shun Suzuki
 * -----
 * Last Modified: 08/03/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2022 Shun Suzuki. All rights reserved.
 * 
 */


using AUTD3Sharp;
using AUTD3Sharp.Gain;
using AUTD3Sharp.Modulation;
using AUTD3Sharp.Utils;
using Custom = AUTD3Sharp.Gain.Custom;

namespace Samples;
internal static class CustomTest
{
    private static Gain Focus(Controller autd, Vector3d point)
    {
        var amps = new double[autd.Geometry.NumTransducers];
        var phases = new double[autd.Geometry.NumTransducers];

        foreach (var tr in autd.Geometry)
        {
            var tp = tr.Position;
            var dist = (tp - point).L2Norm;
            var wavenum = 2.0 * AUTD3.Pi / tr.Wavelength;
            var phase = dist * wavenum;
            amps[tr.Id] = 1.0;
            phases[tr.Id] = phase;
        }

        return new Custom(amps, phases);
    }

    public static void Test(Controller autd)
    {
        var config = new SilencerConfig();
        autd.Send(config, TimeSpan.FromMilliseconds(20));

        var mod = new Sine(150);
        var gain = Focus(autd, autd.Geometry.Center + new Vector3d(0, 0, 150));
        autd.Send(mod, gain, TimeSpan.FromMilliseconds(20));
    }
}
